Update install docs to clarify install steps for developers vs. users

brianhelba opened the pull request:
The Grunt CLI does not directly depend on the Grunt task runner.

See http://gruntjs.com/getting-started#installing-the-cli

zachmullen commented:
@brianhelba we install the grunt worker locally via package.json, is this actually necessary?

jbeezley commented:
I think when you install locally then you have to run grunt with the local path `./node_modules/.bin/grunt`.

zachmullen commented:
The CLI is installed globally, but the linked docs say that the CLI uses the grunt installed next to the Gruntfile being executed.

zachmullen commented:
```
[zach girder]$ rm -rf node_modules/grunt
[zach girder]$ grunt
grunt-cli: The grunt command line interface. (v0.1.13)

Fatal error: Unable to find local grunt.

If you're seeing this message, either a Gruntfile wasn't found or grunt
hasn't been installed locally to your project. For more information about
installing and configuring grunt, please see the Getting Started guide:

http://gruntjs.com/getting-started
```

jbeezley commented:
I'm not sure what the issue is here, but as I recall, you **do** need to install `grunt` locally either way.  You **don't** need `grunt-cli` installed locally though.

brianhelba commented:
@zachmullen You're right, I've looked closer in the "package.json", and since it lists Grunt / Grunt CLI as a dependency and as a post-install step, we can just remove all direct references to Grunt as an explicit package to install and as an explicit build step to run.

That is, a user just needs to run "npm install" from inside the root "girder/" directory. All other Grunt-related steps should happen automatically at that point.

Not only is this simpler for users (removing 2 installation steps, and making the build always be just 1 step), but it makes Grunt / Grunt CLI both get installed locally, which is the best practice http://blog.nodejs.org/2011/03/23/npm-1-0-global-vs-local-installation .

zachmullen commented:
Agreed, though I'm hesitant to remove the instruction about installing the CLI globally, since elsewhere in the docs we frequently assume "grunt" is on the PATH, and we have existing users who also are used to that being the case. What do others think? @girder/developers @girder/owners

jbeezley commented:
:+1: To installing `grunt-cli` globally.

brianhelba commented:
@zachmullen @jbeezley So, I've updated my PR to remove this entirely, because I think it's probably the "right" solution. I'm fine to not merge this yet, and continue discussion. We can go in another direction if that's the consensus.

cpatrick commented:
Part of me is okay with continuing the grunt-cli global install. The other part of me thinks that we should add additional npm targets for the things that need the grunt cli globally later.

brianhelba commented:
So here's an interesting compromise, the "npm install" command can be run with "-g", to install required packages globally (though that would include everything, not just grunt). https://docs.npmjs.com/cli/install

brianhelba commented:
Additionally, note that http://blog.nodejs.org/2011/03/23/npm-1-0-global-vs-local-installation reccomends that we install Grunt twice, once globally for developers, and once locally to run the app with.

We could update the docs so that the standard install procedure does not mention Grunt (as my PR does). Then add a "Development Setup" section, instructing users to install Grunt globally, install "requirements-dev.txt" with pip, install CMake, etc.

jbeezley commented:
@cpatrick npm targets?

@brianhelba `npm install -g grunt-cli` is okay, but `npm install -g` for all dependencies is a bad idea.  I have seriously messed up my global package repository doing that before.

zachmullen commented:
> npm install -g grunt-cli is okay, but npm install -g for all dependencies is a bad idea. I have seriously messed up my global package repository doing that before.

Agreed. Then we are potentially clashing with any other projects using npm packages on the system, and I don't want to be a bad citizen by polluting global scope.

> We could update the docs so that the standard install procedure does not mention Grunt (as my PR does). Then add a "Development Setup" section, instructing users to install Grunt globally, install "requirements-dev.txt" with pip, install CMake, etc.

I'd be OK with that if others are.

jbeezley commented:
Sounds good to me.

brianhelba commented:
@jbeezley I now agree, "npm install -g" is a bad idea, as I just tested it, and it does some wacky things.

For everyone, I'm now more convinced that all references to Grunt and installing Grunt globally should be moved to the development documentation page.

I see the biggest advantage here in simplifying the installation / build process for new users, particularly those who are naive to a JS-based build environment. I can testify first-hand that when I first installed Girder, I was not familiar with the details of NPM or Grunt, and I was a little overwhelmed at how many new technologies there were. The more we can abstract things away, the better.

I'm updating my PR now.

cpatrick commented:
@jbeezley I assumed that [scripts](https://docs.npmjs.com/files/package.json#scripts) was more versatile than it actually is...

brianhelba commented:
@zachmullen @jbeezley @cpatrick 
I've rebased this and added another commit to address the issues that have been discussed so far.

This is still WIP, as I want to add a bit more cleanup:
* clarify the prerequisites for building documentation
* add install instructions for CMake; none of the global package install instructions in "prerequisites.rst" include it; that said, some slower distros may not be able to use the package manager to get CMake, as the packaged version might be too old
* note that MongoDB does not *have* to be installed locally; users can use connect to a MongoDB server running somewhere else
* if they are any different, separate the system prerequisites for running the server vs. development

zachmullen commented:
:+1:

jbeezley commented:
Looking good.

jcfr commented:
:+1:

zachmullen commented:
@brianhelba is this still WIP?

zachmullen commented:
@brianhelba I'd like to merge this since it contains good changes, but it's still labeled as WIP. Can we get this merged and defer your other proposed changes to a future branch?

brianhelba commented:
@zachmullen Yeah, lets do that.

I'd hoped to have a little time to update the rest of the items, but I've been so busy.

brianhelba commented:
Rebased. I'll merge when the tests come back.
